Updates to API Puller, Network Check, unsubmerged sonde flag AND add_column_if_not_exists

SamStruthers opened the pull request:
Hi @juandlt-csu & @kathryn-willi 

I was using the package to update our 2025 API pull and had to create a few custom versions of functions where this package fell short. 
# Changes:
## `api_puller`:
By default, FDOM is always excluded from any API pull. I added an argument `network` so that the FCW system (default) still filters FDOM out but for all other uses, FDOM is allowed to pass through
### Review:
- pretty minimal change, so just checking to make sure this wouldn't affect the FC IT system would be good

## `network_check`:
 I played around with a solution to #9 where users can specify a specific site order function `site_order_fun` that takes in a site name and returns a site_order vector or NULL (no network check needed). If the network is "fcw" (default), we still follow the expected order for PWQN sites.  It worked on a local testing version I had for the 2025 data but curious to hear your thoughts on this method. I added an example of this to the vignette (Step 9)
### Review:
- same as api_pull but thoughts on having a function as an input. 

## `add_unsubmerged_flag`:
 A possible solution to #16. Added a argument, `method`, so users can specify if they want to check for this flag with either depth (`depth`, default) or sc (`sc`)or both (`depth_sc`). I set the SC boundary to be quite low (<= 10) to be pretty conservative/low since we have some upper sites ("sfm") where SC is quite low in general (~25 ). We could/should revisit this with all the historical data to see if we actually catch any dry sondes with this added method. 
### Review:
- Check logic to make sure I am not introducing any unseen errors with joins/filtering
- Thoughts on allowing users to define the method

## EDIT:
## `add_column_if_not_exists`:
Took out this function used in multiple places (`fix_turbidity`,`network_check`, `low_pass_filter` & `generate_summary_statistics` ) and made it into it's own function. 
### Review:
- Double check this doesn't break our usage in other functions
- General typos in documentation?

kathryn-willi commented on vignettes/using_the_package_in_rstudio.Rmd:
```diff
@@ -375,11 +375,59 @@ iwalk(intrasensor_flags_list,
+                     "archery")
+    return(sites_order)
+  }
+  # disconnected/independant sites: mtn campus, sfm
```
```suggestion
  # disconnected/independent sites: mtn campus, sfm
```

kathryn-willi commented on R/network_check.R:
```diff
@@ -20,16 +20,19 @@
 #' intra-sensor flagging functions which are indexed by their corresponding 
 #' site-parameter combination.
-#' @param network Whether to perform the network crawl across FCW sites only ("fcw") or
+#' @param network Whether to perform the network crawl across FCW sites only ("fcw", Default) or
```
```suggestion
#' @param network Whether to perform the network crawl across FCW sites only ("fcw", Default) or a "custom" site list
```

kathryn-willi commented on R/network_check.R:
```diff
@@ -20,16 +20,19 @@
 #' site-parameter combination.
-#' @param network Whether to perform the network crawl across FCW sites only ("fcw") or
+#' @param network Whether to perform the network crawl across FCW sites only ("fcw", Default) or
 #' all sites ("all")
```
```suggestion

```

based on the code in the vignette, the "all" option no longer exists, right?

kathryn-willi commented on R/add_unsubmerged_flag.R:
looks good

kathryn-willi commented on R/api_puller.R:
looks good

juandlt-csu commented on R/add_unsubmerged_flag.R:
```diff
@@ -3,40 +3,89 @@
+    }
+    
+  } else {
+    stop("Invalid method. Choose from 'depth_sc', 'depth', or 'sc'.")
```
Future to do: These argument checks should be pulled out of this `ifelse` nest and be done at the top of the function.

juandlt-csu commented on R/add_unsubmerged_flag.R:
Looks good to me, I would just like to see the argument checks at the top of the function to avoid large `if else` nests. Makes future debugging and developing much easier.

juandlt-csu commented on R/network_check.R:
```diff
@@ -47,38 +50,18 @@ network_check <- function(df, network = "all", intrasensor_flags_arg = intrasens
-    # SFM network 
-    if (site_name %in% c("mtncampus", "sfm")){
-      sites_order <- c("mtncampus", "sfm")
+    #use site_order_fun to get the order of sites if not in CSU or FCW network
+    sites_order <- site_order_fun(site_name)
+    # If site is not in a defined order, return original df
+    if (is.null(sites_order)) {
+      return(df)  # no site order, therefore no network check
     }
```
This behavior is confusing and not explained in the documentation. I think the documentation should explain the interactions between arguments if one of those arguments is itself a function. If that function is not defined in `R/` then there should be some sort of example that is easy to follow in the docs.

For example, the assumption here is that the custom `site_order_fun` should always have a `NULL` return option available instead of something else, and that that option should always mean the same thing (across different custom functions), so that this function returns consistent outputs.

juandlt-csu commented on R/network_check.R:
I think that there are a couple of issues with the updates to this function that should be fixed and revisited. These updates create more complex interactions than are accounted for in the documentation and the vignette. 

At the very least the comments, docs and vignette need to be updated so that the interactions between objects and assumptions of those objects are documented accurately. 

More fundamentally, I think that the `site_order_fun` arg should be revisited as the solution for a more fluid site order list generator. I love the underlying idea, and I think that the implementation here was clever, but I also think that it is confusing to follow and would be hard to maintain/update in the future. Some of this could be fixed with better documentation, but I think that this should just be re-worked.

Here is an idea of what I think would work:
- `network_check` was designed to take in a site-parameter df as input and be used in a map. 
- The desire is that the `site_order` object be fluid and not hard coded, but when it is mapped those site order lists are expected to be static.
- Instead of a custom function as an input to decide the site order, a list of site orders should be supplied to the function as an argument. These are static objects, and the custom function was just used to decide which of the lists to use. That aspect should just be in network check since it's just checking in which list the name falls in. 
- IMO this means that instead of a fluid, non documented custom function, another function should be made that makes these static lists to supply into network check. I would make a yaml parsing function which makes lists from a yaml consistent yaml structure. 
- This makes it so that that yaml parsing function is static, consistent, generalized, and well documented. It also makes it possible to document a consisted structure for the custom yaml file, which are easier to make than functions for users and are less ephemeral than functions in scripts.

This is not me saying that you need to do this^ exactly, but given that this package is currently where it seems like we are going to base our long term functions out of, I think that we should approach these updates with a more maintainable approach in mind.

Some other things that I noticed:
- The updates of this function make the internal `sites_order` object potentially very fluid, but there is no indication of this anywhere in the update. 
- This function is mapped. Every time it is mapped it is running the `site_order_fun` to re generate the lists, but those are static across all the maps. This should be some static input instead. 
- The potential inputs for the args should be paired down and made explicit in the code. (i.e. allow users to pick "all", "lower", "upper", or "custom", with potentially other options for virridy sites etc)
- The docs state that some parameters are handled differently (lines 10-12) but there is no way to interact with that outside of the function. this should be revisited. Not necessarily a task for this PR, but in the future.

juandlt-csu commented on vignettes/using_the_package_in_rstudio.Rmd:
```diff
@@ -375,11 +375,59 @@ iwalk(intrasensor_flags_list,
   # network check compares patterns across sites
-  purrr::map(~network_check(df = ., intrasensor_flags_arg = intrasensor_flags_list)) %>%
+  purrr::map(~network_check(df = ., intrasensor_flags_arg = intrasensor_flags_list,
+                            network = "custom", site_order_fun = my_site_order)) %>% # define network as custom and input a function to define site order by site name
```
Related comments are in my review for `network_check`

juandlt-csu reviewed (changes requested):
Hey Sam, I like a lot of the changes that were made here, but I think that the updates to `network_check` would be too hard to maintain in the long run so I am requesting that some changes be made to that. At least some more robust documentation is needed, if not a re-work of that idea. I am okay with this being added with robust documentation and the creation of an issue which addresses those problems that I brought up in my review so that it can be addressed in the future if the updates in this PR need to be added ASAP or if you don't have the time/energy to make those updates right now.

SamStruthers commented on R/api_puller.R:
Love this and will change accordingly! I was also a bit confused why this was in here but c'est la vie!

SamStruthers commented:
Hi @juandlt-csu, 
Thanks for all the notes on `network_check()`, I think I was rushing a bit and did not create the best version of the function. 
A few changes:
## api_puller:
- Removed the hardcoded FDOM filter entirely
## Munge_api_data:
- Removed Vulink and virridy filters which were hardcoded, we should do these manually if we want or setup a new argument to decide when to do this

## network_check/load_site_order: 
- Per our meeting, updated the site order yaml/csv files (see inst/templates)
- I then updated `load_sites_order()` to match this new logic
- And updated network check to select the correct `sites_order` based on site_name
I think the could be a couple of edge cases where the `sites_order` logic could fail if other users had **very** different use cases but for all our work, the logic works correctly

## add_unsubmerged_flag:
- moved if_else statement to top

## using_in_rstudio
- Added in munge_api_data call and showed how to filter out the data we don't want in the FC system
- Showed how to pull in site order lists and use in network check

juandlt-csu commented on R/load_site_order.R:
```diff
@@ -0,0 +1,104 @@
+    # Convert dataframe to named list
+    site_order <- site_order_raw %>%
+      #convert network string into separated, trimmed character vectors
+      dplyr::mutate(network = stringr::str_split(network, ",")) %>%
```
Is the preference for a string containing the site order information over a vector based on ease of use for users who are not code-savvy?

juandlt-csu commented on R/network_check.R:
```diff
@@ -20,64 +20,47 @@
+  if(length(sites_order) > 1 & (site_name %in% names(sites_order))){
+    sites_order <- sites_order[[which(names(sites_order) == site_name)]]
+  }else{
+    sites_order <- sites_order[[1]]
```
Clarification: We are going to base the site list decision based on the order of the lists for now?

juandlt-csu commented on R/network_check.R:
```diff
@@ -20,64 +20,47 @@
+  }
+  
+  # If no site order is found for the site (ie only one site in network list), return original dataframe with message
+  if(length(sites_order) ==  1){
+    message(paste0("Skipping network check (no upstream/downstream relationship) for: ", site_name))
+    return(df)
```
I wonder if in the future we would just add comments in the yaml file saying which sites don't have relationships, so that L49 would trigger for those instead. The explicit nature of the yaml right now is fine, this is just a thought.

juandlt-csu reviewed (approved):
Thanks for making these changes @SamStruthers! I made some very minor notes, but am very happy with this, go ahead and merge!

kathryn-willi commented on R/api_puller.R:
```diff
@@ -80,10 +80,8 @@ api_puller <- function(site,
-    # For FCW/CSU networks exclude FDOM parameter
-    dplyr::filter(parameter != "FDOM Fluorescence")
-  
+    dplyr::select(site, id, name, timestamp, parameter, value, units) 
```
This removal shouldn't in theory be a problem for FCW, since only Virridy sondes have FDOM right? We definitely want to make sure no FDOM data gets put into the FCW system.

SamStruthers replied:
Our sondes (everywhere except Bellvue) have FDOM. It would be easy to filter out at any point though so we just need to double check that we do (or that the city's contractors know that FDOM is not for the public)

kathryn-willi commented on R/network_check.R:
```diff
@@ -20,64 +20,47 @@
+#' @seealso [load_site_order()]
 
-network_check <- function(df, network = "all", intrasensor_flags_arg = intrasensor_flags) {
+network_check <- function(df, intrasensor_flags_arg = intrasensor_flags, site_order_arg = site_order_list){
```
This will break the FCW code, just a heads up to make changes to their system once this gets included to their system!

juandlt-csu replied:
We will have to add the `site_order_list` information for sure.

kathryn-willi commented on R/munge_api_data.R:
```diff
@@ -73,13 +74,13 @@ munge_api_data <- function(api_dir, summarize_interval = "15 minutes",
+    #dplyr::filter(!grepl("vulink", name, ignore.case = TRUE)) %>%
     # Filter out Virridy data (not used in CSU/FCW networks)
-    dplyr::filter(!grepl("virridy", name, ignore.case = TRUE)) %>%
+    #dplyr::filter(!grepl("virridy", name, ignore.case = TRUE)) %>%
```
this will also break the code in FCW - just as a heads up for when we make changes inside the system. The Virridy data should never be stored in the FC system. (In fact, we probably want to get a list of site names ahead of downloading the data so that their system never touches Virridy data)

juandlt-csu replied:
Yeah, that already happens in the synapse code:

```
hv_sites <- hv_locations_all(hv_token) %>%
  # Filter out VuLink data (not used in CSU/FCW networks)
  dplyr::filter(!grepl("vulink", name, ignore.case = TRUE),
                # Filter out Virridy sondes (not part of CSU/FCW networks)
                !grepl("virridy", name, ignore.case = TRUE),
                # Filter for sites that are in our site list
                grepl(sites_str, name, ignore.case = TRUE))
```

As far as parameters go, that is also taken care of in the fc script:
```
# Define the standard set of water quality parameters to process
# This ensures consistent processing even if some parameters are missing at certain sites
params <- c(
  "Chl-a Fluorescence",  # Chlorophyll-a, indicator of algal concentration
  "Depth",               # Water depth at sensor location
  "DO",                  # Dissolved oxygen
  "ORP",                 # Oxidation-reduction potential
  "pH",                  # Acidity/alkalinity
  "Specific Conductivity", # Conductivity normalized to 25°C
  "Temperature",         # Water temperature
  "Turbidity"            # Water clarity/suspended particles
)

# Create a Comprehensive List of Site-Parameter Combinations
# ========================================================
# Generate all possible combinations of sites and parameters
# This creates a reference list for processing that includes all
# theoretically possible site-parameter pairs
site_param_combos <- tidyr::crossing(sites, params) %>% 
  dplyr::mutate(combo = paste0(sites, "-", params)) %>% 
  dplyr::pull(combo) 

# Filter the new data to only include the relevant site-parameter combinations
# ========================================================= 
new_data_subset <- new_data[names(new_data) %in% site_param_combos]
```

So we prevent the virridy/vulink data from ever touching the fc system. We can't prevent FDOM touching the system, but we do filter it out. I will double check to make sure that this behaves as expected during the winter changes to see where things break if they do.

SamStruthers replied:
Great! Thanks for checking on that
